store/tooling/tooling.go

```diff
@@ -437,3 +531,64 @@ func (tsto *ToolingStore) AssertionMaxFormats() map[string]int {
 func MockToolingStore(sto StoreImpl) *ToolingStore {
 	return &ToolingStore{sto: sto}
 }
+
+type DownloadedComponent struct {
+	Path string
+	Info *snap.ComponentInfo
+}
+
+func (tsto *ToolingStore) componentDownload(targetFn string, snapName string, srr *store.SnapResourceResult, opts DownloadSnapOptions) (downloadedComp *DownloadedComponent, err error) {
+	// Check if this is a component we can handle
+	ctyp, err := store.ResourceToComponentType(srr.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// check if we already have the right file
+	if osutil.FileExists(targetFn) {
+		sha3_384Dgst, size, err := osutil.FileDigest(targetFn, crypto.SHA3_384)
+		if err == nil && size == uint64(srr.DownloadInfo.Size) &&
+			fmt.Sprintf("%x", sha3_384Dgst) == srr.DownloadInfo.Sha3_384 {
+
+			logger.Debugf("not downloading, using existing file %s", targetFn)
+			return &DownloadedComponent{
+				Path: targetFn,
+			}, nil
+		}
+		logger.Debugf("File exists but has wrong hash, ignoring (here).")
+	}
+
+	pb := progress.MakeProgressBar(tsto.Stdout)
+	defer pb.Finished()
+
+	// Intercept sigint
+	c := make(chan os.Signal, 3)
+	signal.Notify(c, syscall.SIGINT)
+	go func() {
+		<-c
+		pb.Finished()
+		os.Exit(1)
+	}()
+
+	dlOpts := &store.DownloadOptions{LeavePartialOnError: opts.LeavePartialOnError}
+	cref := naming.NewComponentRef(snapName, srr.Name)
+	if err = tsto.sto.Download(context.TODO(), cref.String(), targetFn,
+		&srr.DownloadInfo, pb, nil, dlOpts); err != nil {
+		return nil, err
+	}
+
+	signal.Reset(syscall.SIGINT)
```

Comment:
nitpick: wonder if we should extract  the ProbressBar and Signal logic that is the same for snaps in a small helper that takes a closure/function ?

Reply:
Good idea, I've done this now